Openresty

julek-wolfssl opened the pull request:
This PR is branched off of #3646 so it has to wait for that PR to merge.

haydenroche5 reviewed (approved):
Compiles successfully for me.

julek-wolfssl commented:
Retest this please.

SparkiDev commented on src/ssl.c:
```diff
@@ -15797,6 +15900,21 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
+            WOLFSSL_MSG("Bad parameter");
+            return WOLFSSL_FAILURE;
+        }
+        WOLFSSL_MSG("WARNING: This function modifies the cert store used by all "
```
This looks dangerous. Openresty uses the same store for all SSL objects?

julek-wolfssl replied:
It probably is. I'm guessing its better to guard it with some kind of an ALLOW_UNSAFE macro.

SparkiDev replied:
Todd and I agree that adding a pointer to the store in the SSL would be fine.
Please do this.

julek-wolfssl replied:
How should I change the existing references to `ssl->ctx->cm` and others to use the WOLFSSL store? I was thinking of defining macros like:
```
#define SSL_GET_STORE(ssl) (ssl->x509_store_pt ? ssl->x509_store_pt : ssl->ctx->x509_store_pt)
#define SSL_GET_CM(ssl) (ssl->x509_store_pt ? ssl->x509_store_pt->cm : ssl->ctx->cm)
```
What do you think of this idea?

SparkiDev replied:
The pointer to the store (cm) will be copied into the WOLFSSL object on wolfSSL_new().
If you need to modify the store locally, then duplicate it and mark for free.
Make sense?

julek-wolfssl replied:
I'm not sure if copying the cm is a good idea. This would break existing connections if the CTX store is set to a new value. If we would up the refCount for every SSL object then a new cm in the CTX would still not be registered in the SSL objects. If you think that this is acceptable behaviour then I will go ahead and change it.

SparkiDev replied:
Sorry. Forgot what the API did.
Yeah just replace the pointer.

julek-wolfssl replied:
So using the macros I suggested earlier? I'm sorry I don't understand what you mean by "just replace the pointer".

SparkiDev replied:
Why macro?
Copy the pointer down in the InitSSL.

julek-wolfssl replied:
Ok. And just to clarify, it is ok if a new cm is set in the ctx and this is not registered by existing ssl objects? I'm asking because this changes how the cm functions currently.

SparkiDev replied:
After talking to Todd, he would be happy with up reffing (incrementing reference count) on the CertManager.
The WOLFSSL_X509_STORE may need to have have a reference count as well.
Use a  macro as you have above. That is, if the SSL object doesn't have a store then use the context's store.
If the store being set is the one in the context then don't do anything.

julek-wolfssl replied:
Thanks for the clarification Sean! This clears everything up.

julek-wolfssl replied:
In the last commit (a5a5b3d4678a3e3408132066118ca02305aefd65) I added a `WOLFSSL_X509_STORE` to the SSL object, added a ref count, and substituted calls with macros choosing the appropriate object.

SparkiDev commented on src/ssl.c:
```diff
@@ -44016,6 +44248,32 @@ int wolfSSL_i2d_X509_NAME(WOLFSSL_X509_NAME* name, unsigned char** out)
+                    "SSL objects created using the same CTX");
+        if (ssl == NULL || ssl->ctx == NULL)
+            return WOLFSSL_FAILURE;
+        return wolfSSL_CTX_add1_chain_cert(ssl->ctx, x509);
```
Do we actually want to change the context?

julek-wolfssl replied:
We don't currently support per SSL chains. I can look into how difficult this would be. Alternatively, this could be a stub that doesn't do anything/errors out.

SparkiDev replied:
Please make a change here or give an explanation as to why not.

julek-wolfssl replied:
Changed in https://github.com/wolfSSL/wolfssl/pull/3726/commits/f440b406c147772e9d7b17e5ebad987b67bd0e84 to not affect the context.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15865,6 +15970,53 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
+        }
+
+        /* NO-OP when setting existing store */
+        if (str == ssl->ctx->x509_store_pt || str == ssl->x509_store_pt)
```
Modified to take that into account. In this case we revert to using the context store.

SparkiDev commented on src/tls13.c:
```diff
@@ -8458,24 +8467,34 @@ int wolfSSL_accept_TLSv13(WOLFSSL* ssl)
+                !ssl->buffers.certificate->buffer) {
+
+                WOLFSSL_MSG("accept error: server cert required");
+                WOLFSSL_ERROR(ssl->error = NO_PRIVATE_KEY);
```
WOLFSSL_ERROR as a macro may map to nothing and this assignment won't happen.
Separate out assignment from macro call.

julek-wolfssl replied:
Separated out.

SparkiDev commented on wolfssl/ssl.h:
```diff
@@ -553,9 +553,11 @@ struct WOLFSSL_X509_STORE {
+#ifdef HAVE_CRL
     WOLFSSL_X509_CRL *crl; /* points to cm->crl */
 #endif
+    wolfSSL_Mutex   refMutex;   /* reference count mutex */
```
Limit the inclusion of these fields.
Specifically SINGLE_THREADED.

julek-wolfssl replied:
Limited all `refMutex` fields I could find with `SINGLE_THREADED`.

SparkiDev commented on src/ssl.c:
```diff
@@ -56217,6 +56476,216 @@ int wolfSSL_X509_CA_num(WOLFSSL_X509_STORE* store)
 }
 
+#ifndef NO_WOLFSSL_STUB
+int wolfSSL_X509_STORE_get_by_subject(WOLFSSL_X509_STORE_CTX* ctx, int idx,
```
These functions have X509_STORE in the API but if you look at the parameters, they operate on a WOLFSSL_X509_STORE_CTX. Unless the API definition is wrong, put it back with the other functions that operate on a WOLFSSL_X509_STORE_CTX.

julek-wolfssl replied:
This is because OpenSSL has this API defined as:
`# define X509_STORE_get_by_subject X509_STORE_CTX_get_by_subject`
Moved back.

SparkiDev replied:
Thanks.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -43695,18 +43927,154 @@ int wolfSSL_i2d_X509_NAME(WOLFSSL_X509_NAME* name, unsigned char** out)
+     * wolfSSL_CTX_add1_chain_cert makes a copy of the cert so we free it
+     * on success
+     */
+    int wolfSSL_CTX_add0_chain_cert(WOLFSSL_CTX* ctx, WOLFSSL_X509* x509)
```
I think that it was easier to do it this way with the call to `wolfSSL_CTX_load_verify_buffer`.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -43695,18 +43927,154 @@ int wolfSSL_i2d_X509_NAME(WOLFSSL_X509_NAME* name, unsigned char** out)
         return WOLFSSL_SUCCESS;
     }
 
+    static int pushCertToDerBuffer(DerBuffer** inOutDer, int weOwn,
```
Changed.

SparkiDev replied:
Also change: pushCAx509Chain

julek-wolfssl replied:
Updated `pushCAx509Chain`.

julek-wolfssl commented on wolfssl/internal.h:
```diff
@@ -4489,6 +4500,21 @@ struct WOLFSSL {
+ * logic for choosing which WOLFSSL_CERT_MANAGER and WOLFSSL_X509_STORE to use.
+ * Always use SSL specific objects when available and revert to CTX otherwise.
+ */
+#if defined(OPENSSL_EXTRA) || defined(HAVE_WEBSERVER) || defined(WOLFSSL_WPAS_SMALL)
```
Added a `WOLFSSL_LOCAL_X509_STORE` macro which is defined in `configure.ac`.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15957,7 +15979,49 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
+     * @param ssl
+     * @return 0 on success
+     */
+    int callCertSetupCb(WOLFSSL* ssl)
```
Changed.

SparkiDev replied:
Thanks

kaleb-himes commented:
retest this please

dgarske commented on src/ssl.c:
```diff
@@ -18838,13 +19028,13 @@ WOLF_STACK_OF(WOLFSSL_X509)* wolfSSL_set_peer_cert_chain(WOLFSSL* ssl)
         }
 #else
-        (void)pushCAx509Chain;
+        (void)PushCAx509Chain;
```
Please fix, so `PushCAx509Chain` is not compiled in when `WOLFSSL_QT`. I would prefer the function not be compiled vs using a `(void)` hack to avoid the unused function. Also not sure why QT would do this?

julek-wolfssl replied:
This is changed in https://github.com/wolfSSL/wolfssl/pull/4059. I'm not sure why this is compiled out for QT.

julek-wolfssl commented on wolfssl/internal.h:
```diff
@@ -4489,6 +4502,19 @@ struct WOLFSSL {
+                  (ssl->ctx->x509_store_pt ? ssl->ctx->x509_store_pt : \
+                                            &ssl->ctx->x509_store))
+#else
+#define SSL_CM(ssl) ssl->ctx->cm
```
That's what I thought as well but `x509_store_pt` is only defined with `WOLFSSL_LOCAL_X509_STORE` so there is no point in defining `SSL_STORE` in this case.

dgarske replied:
I searched a saw a place `SSL_STORE ` could be used. It is still a good idea for future protections too. Please add a `SSL_STORE` macro for the else case.

julek-wolfssl replied:
Do you mean the `LoadCertByIssuer` call in `ProcessPeerCerts`? Also changed in `wolfSSL_set0_verify_cert_store` and `wolfSSL_set1_verify_cert_store`.

I still don't see why we should define `SSL_STORE` in the `#else` case. When `WOLFSSL_LOCAL_X509_STORE` is not defined then no `x509_store*` members are defined in `WOLFSSL` or `WOLFSSL_CTX`. I think its safer and easier to debug to have wolfSSL error out during compilation than during runtime (in the case of just `#define SSL_STORE(ssl) NULL`). Especially since this is an internal macro.

julek-wolfssl commented on wolfssl/internal.h:
```diff
@@ -4474,6 +4486,7 @@ struct WOLFSSL {
 #if defined(OPENSSL_EXTRA)
     WOLFSSL_STACK* supportedCiphers; /* Used in wolfSSL_get_ciphers_compat */
     WOLFSSL_STACK* peerCertChain;    /* Used in wolfSSL_get_peer_cert_chain */
+    WOLFSSL_STACK* ourCertChain;    /* Used in wolfSSL_add1_chain_cert */
```
Added guard.

julek-wolfssl commented on wolfssl/ssl.h:
```diff
@@ -1788,6 +1801,11 @@ typedef int (*client_cert_cb)(WOLFSSL *ssl, WOLFSSL_X509 **x509,
+typedef int (*CertSetupCallback)(WOLFSSL*, void*);
+WOLFSSL_API void wolfSSL_CTX_set_cert_cb(WOLFSSL_CTX* ctx,
+    CertSetupCallback cb, void *arg);
+int CertSetupCbWrapper(WOLFSSL* ssl);
```
Yes. Fixed.

julek-wolfssl commented on wolfssl/ssl.h:
```diff
@@ -103,6 +103,11 @@
 #endif
 
+
+#if defined(OPENSSL_EXTRA) || defined(HAVE_WEBSERVER) || defined(WOLFSSL_WPAS_SMALL)
```
Changed.

julek-wolfssl commented on wolfssl/openssl/ssl.h:
```diff
@@ -820,11 +821,17 @@ wolfSSL_X509_STORE_set_verify_cb((WOLFSSL_X509_STORE *)(s), (WOLFSSL_X509_STORE_
 #define SSL_get_ex_data                 wolfSSL_get_ex_data
 
+#ifndef WOLFSSL_NO_STUB
+#define SSL_set_client_CA_list(...)
```
Currently this can only be controlled at the context level. This could be implemented as part of a future port that needs per-SSL object CA-list definition.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -56663,7 +57092,7 @@ int wolfSSL_X509_STORE_CTX_init(WOLFSSL_X509_STORE_CTX* ctx,
             ctx->current_cert = wolfSSL_X509_d2i(NULL, x509->derCert->buffer,x509->derCert->length);
             if(ctx->current_cert == NULL)
-                return WOLFSSL_FATAL_ERROR;
+                return WOLFSSL_FAILURE;
```
Fixed.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -43717,6 +44085,35 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
+{
+    WOLFSSL_ENTER("wolfSSL_certs_clear()");
+
+    if (ssl == NULL || ssl->ctx == NULL)
```
Not used. Removing check.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -42440,19 +42672,155 @@ int wolfSSL_i2d_X509_NAME(WOLFSSL_X509_NAME* name, unsigned char** out)
+            /* Store cert to free it later */
+            if (ret == WOLFSSL_SUCCESS) {
+                wolfSSL_X509_free(ssl->ourCert);
+                ssl->ourCert = x509;
```
Yes it does. Also need to check this flag before free'ing.

dgarske commented on configure.ac:
```diff
@@ -818,13 +819,25 @@ AC_ARG_ENABLE([openvpn],
 
+if test "$ENABLED_OPENRESTY" = "yes"
+then
+    ENABLED_NGINX="yes"
```
So the openresty also enables NGINX because of the compatibility API's it enables? Is there another way we can group these by openssl feature block or version?

dgarske commented on src/internal.c:
```diff
@@ -3762,10 +3763,12 @@ void InitX509(WOLFSSL_X509* x509, int dynamicFlag, void* heap)
     InitX509Name(&x509->subject, 0, heap);
     x509->dynamicMemory  = (byte)dynamicFlag;
-    #if defined(OPENSSL_EXTRA_X509_SMALL) || defined(OPENSSL_EXTRA)
+    #if defined(OPENSSL_EXTRA) || defined(OPENSSL_ALL)
```
So this section is not available for `OPENSSL_EXTRA_X509_SMALL` now? Is that on purpose?

dgarske commented on src/ssl.c:
```diff
@@ -15996,7 +16018,49 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
+    {
+        int ret = 0;
+        if (ssl->ctx->certSetupCb != NULL) {
+            WOLFSSL_ENTER("CertSetupCbWrapper");
```
Should ENTER be moved above the callback NULL check or maybe just removed/combined with MSG?

julek-wolfssl replied:
Removing `WOLFSSL_ENTER`. Leaving just the `WOLFSSL_MSG`.

dgarske commented on src/ssl.c:
```diff
@@ -51519,24 +51942,30 @@ void wolfSSL_RSA_free(WOLFSSL_RSA* rsa)
         wolfSSL_CRYPTO_cleanup_ex_data(&rsa->ex_data);
 #endif
-#if defined(OPENSSL_EXTRA_X509_SMALL) || defined(OPENSSL_EXTRA)
+#if defined(OPENSSL_EXTRA) || defined(OPENSSL_ALL)
```
We loose `OPENSSL_EXTRA_X509_SMALL` here. Is that changed on purpose?

julek-wolfssl replied:
Reverting back.

dgarske reviewed (changes requested):
Few minor comments. PR is looking good overall.

julek-wolfssl committed:
OpenResty 1.13.6.2 and 1.19.3.1 support

# New or Updated APIs
- wolfSSL_get_tlsext_status_type
- wolfSSL_X509_chain_up_ref
- wolfSSL_get0_verified_chain
- SSL_CTX_set_cert_cb
- SSL_certs_clear
- SSL_add0_chain_cert ssl_cert_add0_chain_cert
- SSL_add1_chain_cert ssl_cert_add1_chain_cert
- sk_X509_NAME_new_null
- SSL_CTX_set_cert_cb
- SSL_set0_verify_cert_store
- SSL_set_client_CA_list

# Other Changes
- Ignore gdbinit
- Add api.c tests for new API
- Add `WOLFSSL_X509_STORE* x509_store_pt` to `WOLFSSL`
- Add macro to select the `WOLFSSL` specific store when available and the associated `WOLFSSL_CTX` store otherwise. Calls to `ssl->ctx->cm` and `ssl->ctx->x509_store*` were replaced by macros.
- NO-OP when setting existing store
- Add reference counter to `WOLFSSL_X509_STORE`
- Cleanup MD5 redundant declarations
- WOLFSSL_ERROR may map to nothing so make assignment outside of it
- refMutex fields are excluded with SINGLE_THREADED macro
- Chain cert refactor
- Make `wolfSSL_add0_chain_cert` and `wolfSSL_add1_chain_cert` not affect the context associated with the SSL object
- `wolfSSL_CTX_add1_chain_cert` now updates the `ctx->certChain` on success and stores the cert in `ctx->x509Chain` for later free'ing

dgarske reviewed (changes requested):
All changes have been reviewed and tested with OSP instructions. Approved, but can you please squash merge this PR? Thanks!

julek-wolfssl commented:
Squashed into one commit.
